ClipboardAPI/clipboard-change-event-explainer.md

```diff
@@ -0,0 +1,210 @@
+## 2. User scenarios
+
+### 2.1 Scenario: Sync clipboard with a remote desktop
+When a user copies text or an image on their local machine, the web-based remote desktop application can detect this clipboard change event through the browser's Clipboard API. Upon detecting the change, the application can automatically send the new clipboard content to the remote desktop environment.
+
+![](img/sync-clipboard-scenario.png)
+
+
+### 2.2 Scenario: Show available paste formats in web based editors
+Web based editors like Excel Online, Word Online may support paste operation in multiple formats. Within the UI, it may show the available formats like csv, image or plain text. The clipboard change event can be used to detect the change in available formats in clipboard and reflect the same on the UI as soon as it is changed. 
+
+#### 2.2.1 Copy multiple cells in a spreadsheet should show multiple paste options
+![](img/paste-format-1.png)
+
+#### 2.2.2 Copy plain text should show only single paste option
+![](img/paste-format-2.png)
+
+Similarly UI elements which depend on clipboard state, like "Paste image from clipboard" in an web based image editor, can be enabled/disabled using the clipboardchange event based on weather correct data format is present in clipboard or not.
+
+## 3. Motivation - Alternative to inefficient polling of clipboard
+Today, a web-app can monitor the system clipboard by polling and reading the clipboard through async clipboard API at regular intervals.
+However, polling is not efficient and this feature aims to introduce an efficient way of notifying web apps when clipboard changes.
+Additionally we must ensure that we monitor the clipboard only when absolutely required, that is, there is at least one document having required permissions and is listening to the clipboard change event. This will be described in design details.
+
+## 4. Example javascript code for detecting clipboard changes:
+
+```javascript
+  function callback(event) {
+      // Read clipboard contents using navigator.clipboard
+      navigator.clipboard.readText().then(text => console.log(text));
+  }
+  // Start listening to the clipboardchange event
+  navigator.clipboard.addEventListener("clipboardchange", callback);
+```
+
+## 5. Event spec details and open questions
+
+### 5.1 User permission requirement
+
+#### 5.1.1 Approach 1 - clipboard-read permission required to listen to clipboardchange event
+Since the clipboard contains privacy-sensitive data, we should protect access to the clipboard change event using a user permission - clipboard-read. The web author should ensure that the site has the permission before it starts listening to this event otherwise the provided event handler won't be invoked whenever the clipboard changes. We should consider logging a warning message if the web author starts listening to clipboardchange without acquiring the permissions since web developers might miss integrating the permissions flow into their user experience.
```

Comment:
"The web author should ensure that the site has the permission before it starts listening to this event otherwise the provided event handler won't be invoked whenever the clipboard changes." How can a web author know if a site has the required permissions or not? Is there an API that can be used to query this user choice?

Reply:
"web developers might miss integrating the permissions flow into their user experience" -> The permissions flow implemented and enforced by the browser, not the site. Right?

Reply:
Yes, there is an API for checking if the current user has clipboard-read permissions for the site:

navigator.permissions.query({ name: "clipboard-read" }) (https://www.w3.org/TR/permissions/#query-method_)

The permissions flow is NOT enforced by the clipboardchange event unless we decide to prompt user for permissions as soon as they start listening to the event.